pkg/storegateway/indexheader/binary_reader.go

```diff
@@ -533,84 +534,79 @@ func newFileBinaryReader(path string, postingOffsetsInMemSampling int, cfg Binar
 		// load the whole offset table into memory.
 		r.postingsV1 = map[string]map[string]index.Range{}
 
 		var prevRng index.Range
-		if err := index.ReadOffsetTable(r.b, r.toc.PostingsOffsetTable, func(key []string, off uint64, _ int) error {
-			if len(key) != 2 {
-				return errors.Errorf("unexpected key length for posting table %d", len(key))
-			}
-
-			if lastKey != nil {
+		if err := readPostingsOffsetTable(r.b, r.toc.PostingsOffsetTable, func(lbl labels.Label, off uint64, _ int) error {
+			if lastSet {
 				prevRng.End = int64(off - crc32.Size)
-				r.postingsV1[lastKey[0]][lastKey[1]] = prevRng
+				r.postingsV1[lastLbl.Name][lastLbl.Value] = prevRng
 			}
 
-			if _, ok := r.postingsV1[key[0]]; !ok {
-				r.postingsV1[key[0]] = map[string]index.Range{}
-				r.postings[key[0]] = nil // Used to get a list of labelnames in places.
+			if _, ok := r.postingsV1[lbl.Name]; !ok {
+				r.postingsV1[lbl.Name] = map[string]index.Range{}
+				r.postings[lbl.Name] = nil // Used to get a list of labelnames in places.
 			}
 
-			lastKey = key
+			lastSet = true
+			lastLbl = lbl
 			prevRng = index.Range{Start: int64(off + postingLengthFieldSize)}
 			return nil
 		}); err != nil {
 			return nil, errors.Wrap(err, "read postings table")
 		}
-		if lastKey != nil {
+		if lastLbl.Name != "" {
 			prevRng.End = r.indexLastPostingEnd - crc32.Size
-			r.postingsV1[lastKey[0]][lastKey[1]] = prevRng
+			r.postingsV1[lastLbl.Name][lastLbl.Value] = prevRng
 		}
 	} else {
 		lastTableOff := 0
 		valueCount := 0
 
 		// For the postings offset table we keep every label name but only every nth
 		// label value (plus the first and last one), to save memory.
-		if err := index.ReadOffsetTable(r.b, r.toc.PostingsOffsetTable, func(key []string, off uint64, tableOff int) error {
-			if len(key) != 2 {
-				return errors.Errorf("unexpected key length for posting table %d", len(key))
-			}
-
-			if _, ok := r.postings[key[0]]; !ok {
+		if err := readPostingsOffsetTable(r.b, r.toc.PostingsOffsetTable, func(lbl labels.Label, off uint64, tableOff int) error {
+			if _, ok := r.postings[lbl.Name]; !ok {
 				// Not seen before label name.
-				r.postings[key[0]] = &postingValueOffsets{}
-				if lastKey != nil {
+				r.postings[lbl.Name] = &postingValueOffsets{}
+				if lastSet {
 					// Always include last value for each label name, unless it was just added in previous iteration based
 					// on valueCount.
 					if (valueCount-1)%postingOffsetsInMemSampling != 0 {
-						r.postings[lastKey[0]].offsets = append(r.postings[lastKey[0]].offsets, postingOffset{value: lastKey[1], tableOff: lastTableOff})
+						r.postings[lastLbl.Name].offsets = append(r.postings[lastLbl.Name].offsets, postingOffset{value: lastLbl.Value, tableOff: lastTableOff})
 					}
-					r.postings[lastKey[0]].lastValOffset = int64(off - crc32.Size)
-					lastKey = nil
+					r.postings[lastLbl.Name].lastValOffset = int64(off - crc32.Size)
+					lastLbl = labels.Label{}
```

Comment:
Why not `lastSet = false`? (Or: why is this even needed? We overwrite `lastSet` and `lastLbl` shortly anyway).

Reply:
Even if not strictly needed, doing `lastSet = false` here may better explain the intention.

Reply:
Instead of adding yet another useless line, I just removed this `lastLbl = labels.Label{}` line.

Once you have a last, you can't _unhave_ it. 

https://github.com/grafana/mimir/pull/3393/commits/30b553389382011be9e2e33f1f9bf7efa2d6774a

Reply:
(Please can you ACK that you're okay with this change so I can merge?)

Reply:
works for me.

Reply:
ACK